File: ai-solutions/04/composer-1/python/main.py
import sys

with open(sys.argv[1], 'r') as f:
    input_data = f.read().strip()

grid = [list(line) for line in input_data.split('\n')]
rows = len(grid)
cols = len(grid[0])

# Directions for 8 neighbors
directions = [
    (-1, -1), (-1, 0), (-1, 1),
    (0, -1),           (0, 1),
    (1, -1),  (1, 0),  (1, 1)
]

def count_adjacent_rolls(r, c):
    count = 0
    for dr, dc in directions:
        nr, nc = r + dr, c + dc
        if 0 <= nr < rows and 0 <= nc < cols and grid[nr][nc] == '@':
            count += 1
    return count

# Part 1: Count accessible rolls
part1 = 0
for r in range(rows):
    for c in range(cols):
        if grid[r][c] == '@' and count_adjacent_rolls(r, c) < 4:
            part1 += 1

# Part 2: Iteratively remove accessible rolls
grid2 = [row[:] for row in grid]
total_removed = 0

while True:
    to_remove = []
    for r in range(rows):
        for c in range(cols):
            if grid2[r][c] == '@':
                adjacent_count = 0
                for dr, dc in directions:
                    nr, nc = r + dr, c + dc
                    if 0 <= nr < rows and 0 <= nc < cols and grid2[nr][nc] == '@':
                        adjacent_count += 1
                if adjacent_count < 4:
                    to_remove.append((r, c))
    
    if not to_remove:
        break
    
    for r, c in to_remove:
        grid2[r][c] = '.'
    total_removed += len(to_remove)

print(part1)
print(total_removed)

File: ai-solutions/04/composer-1/rust/src/main.rs
use std::fs;

const DIRECTIONS: [(i32, i32); 8] = [
    (-1, -1), (-1, 0), (-1, 1),
    (0, -1),           (0, 1),
    (1, -1),  (1, 0),  (1, 1),
];

fn count_adjacent_rolls(grid: &Vec<Vec<char>>, r: usize, c: usize) -> usize {
    let rows = grid.len();
    let cols = grid[0].len();
    let mut count = 0;
    
    for (dr, dc) in DIRECTIONS.iter() {
        let nr = r as i32 + dr;
        let nc = c as i32 + dc;
        if nr >= 0 && nr < rows as i32 && nc >= 0 && nc < cols as i32 {
            if grid[nr as usize][nc as usize] == '@' {
                count += 1;
            }
        }
    }
    count
}

fn main() {
    let input = fs::read_to_string(std::env::args().nth(1).unwrap())
        .unwrap()
        .trim()
        .to_string();
    
    let grid: Vec<Vec<char>> = input
        .lines()
        .map(|line| line.chars().collect())
        .collect();
    
    let rows = grid.len();
    let cols = grid[0].len();
    
    // Part 1: Count accessible rolls
    let mut part1 = 0;
    for r in 0..rows {
        for c in 0..cols {
            if grid[r][c] == '@' && count_adjacent_rolls(&grid, r, c) < 4 {
                part1 += 1;
            }
        }
    }
    
    // Part 2: Iteratively remove accessible rolls
    let mut grid2 = grid.clone();
    let mut total_removed = 0;
    
    loop {
        let mut to_remove = Vec::new();
        for r in 0..rows {
            for c in 0..cols {
                if grid2[r][c] == '@' {
                    let mut adjacent_count = 0;
                    for (dr, dc) in DIRECTIONS.iter() {
                        let nr = r as i32 + dr;
                        let nc = c as i32 + dc;
                        if nr >= 0 && nr < rows as i32 && nc >= 0 && nc < cols as i32 {
                            if grid2[nr as usize][nc as usize] == '@' {
                                adjacent_count += 1;
                            }
                        }
                    }
                    if adjacent_count < 4 {
                        to_remove.push((r, c));
                    }
                }
            }
        }
        
        if to_remove.is_empty() {
            break;
        }
        
        for (r, c) in to_remove.iter() {
            grid2[*r][*c] = '.';
        }
        total_removed += to_remove.len();
    }
    
    println!("{}", part1);
    println!("{}", total_removed);
}

